Lib/test/test_posix.py

```diff
@@ -2194,6 +2194,58 @@ def test_utime(self):
                 os.utime("path", dir_fd=0)
 
 
+class NamespacesTests(unittest.TestCase):
+    """Tests for os.unshare() and os.setns()."""
+
+    @support.requires_subprocess()
+    def subprocess(self, code):
+        import subprocess
+        with subprocess.Popen((sys.executable, '-c', code),
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            encoding="utf-8"
+        ) as p:
+            p.wait()
+            return (
+                p.returncode,
+                tuple(p.stdout),
+                tuple(p.stderr),
+            )
+
+    @unittest.skipUnless(hasattr(os, 'unshare'), 'needs os.unshare()')
+    @unittest.skipUnless(hasattr(os, 'setns'), 'needs os.setns()')
+    @unittest.skipUnless(os.path.exists('/proc/self/ns/uts'), 'need /proc/self/ns/uts')
+    @support.requires_linux_version(3, 0, 0)
+    def test_unshare_setns(self):
+        rc, out, err = self.subprocess("""if 1:
+            import os
+            import sys
+            fd = os.open('/proc/self/ns/uts', os.O_RDONLY)
+            try:
+                print(os.readlink('/proc/self/ns/uts'))
+                os.unshare(os.CLONE_NEWUTS)
+                print(os.readlink('/proc/self/ns/uts'))
+                os.setns(fd, os.CLONE_NEWUTS)
+                print(os.readlink('/proc/self/ns/uts'))
+            except OSError as e:
+                sys.stderr.write(str(e.errno))
+                sys.exit(2)
+            finally:
+                os.close(fd)
+            """)
+
+        if rc == 2:
+            e = int(err[0])
+            self.assertIn(e, (errno.EPERM, errno.EINVAL, errno.ENOSPC, errno.ENOSYS))
+            self.skipTest(f"could not call os.unshare / os.setns [Errno {e}].")
+
+        self.assertEqual(rc, 0)
+        self.assertEqual(err, ())
+        original, new, back = out
+        self.assertNotEqual(original, new)
+        self.assertEqual(original, back)
```

Comment:
I would prefer to implement these tests inside the subprocess directly. Playing with stdout is fragile :-(